src/transformers/generation/utils.py

```diff
@@ -92,6 +93,13 @@
 if is_accelerate_available():
     from accelerate.hooks import AlignDevicesHook, add_hook_to_module
 
+
+class CacheImplementation(str, Enum):
+    DYNAMIC = "dynamic"
+    STATIC = "static"
+    SINK = "sink"
```

Comment:
this prevents anyone from adding / using a custom implementation why not just use strings?

Reply:
I think in general it is a good practice to use enums to avoid silent behaviours / errors. 
E.g. if one passes 
```python
model.generation_config.cache_implementation = "Static"
```
The code will silently work out of the box as it will use dynamic cache, and can potentitally lead to silent errors. 

When using enums, 

```python
model.set_cache_implementation("Static")
```
Will also work out of the box but this time will correctly use the static cache implementation and not dynamic cache as opposed to the snippet above.

We can also use a mapping with hardcoded strings but I found it clearer to have enums 

cc @amyeroberts @gante @tomaarsen wdyt?

Reply:
good practice does not mean it should always be used.
Here it's: 
- useless: we need a mapping with keys from `"string_cls":cls`
- cumbersome: anywhere you compare the generation_config.cache_cls (a string) you need to use the enum. Why? 
- not restrictive, and let me be clear here I do NOT want users to pass anything and expect it to work. If the class is not in the Mapping error out and we are done with it. 

TLDR; why would we use then when we don't need it and it only adds additional calls to `CacheImplementation.DYNAMIC` vs `"dynamic"` is something I don't understand.

Reply:
@ArthurZucker I don't understand your objections here. For one, best practices does mean it should be done wherever possible.

I don't believe using enums is useless or cumbersome (aside from these not being well defined)? As @younesbelkada highlights, they provide better guarantees of selection of valid types. The explicit enum means the user can pass a string, but the code uses stricter checks that string matching, which are error prone. 

> not restrictive, and let me be clear here I do NOT want users to pass anything and expect it to work. If the class is not in the Mapping error out and we are done with it. We should raise and error

Enums are restrictive? In fact, they're far more restrictive than doing string checks and have stronger guarantee's than checking a mapping which is mutable.

Reply:
But that is exactly what we want: the mapping should be mutable, because we want it to be easily adapted for a custom code on the hub. 
Down for stricter checks, but in this specific case IMO it is cumbersome and useless as we don't have the notion of "safety" and we basically use mapping for model_type, config_type etc etc.

Reply:
Anyway it's not that important

Reply:
Ah, OK, I understand better now. I'm not sure how we can both let users modify a dict and have guarantees about not working if users can pass anything. If I've understood correctly, is the check you're wanting on the membership within the dictionary rather than e.g. how it's handled within the setting logic? In this case, I agree dictionaries are probably the simplest solution.

Reply:
Yes exactly 😉